add ChangeNotify and drive support for it

remusb opened the pull request:
Hi @ncw 

I've added a new feature which is similar to DirChangeNotify to fs. The reason for an almost identical feature is to keep the workflows of DirChangeNotify unchanged.

The main difference would be in its implementation. It should return both files and directories once per poll interval and make no assumptions about how would they be needed. 

I've also added support for it in drive. I'll take a look at other cloud providers if they could make use of it.

Let me know what you think. The diff in drive is ugly but I just refactored the code a bit as the drive integration would be identical in both of them.

ncw commented:
> I've added a new feature which is similar to DirChangeNotify to fs. The reason for an almost identical feature is to keep the workflows of DirChangeNotify unchanged.

I'd rather change the callers of DirChangeNotify (there aren't many) than have two very similar implementations.

Could the interface be changed so that the callback includes the path and an indication of whether the called-back thing is a file or a directory?

remusb commented:
Sure. Thanks

remusb commented:
Ok, refactored these interfaces:

- rename DirChangeNotify to ChangeNotify
- added a new type `EntryType` with 2 constants in fs: EntryDirectory and EntryObject
- updated polling functionality in acd and drive to use only ChangeNotify
- both of them now push unique changed paths in a similar way:
   - if it's a file, it will take its first parent, use dir cache to convert it to a path and append the file name
   - if it's a folder, it will send it as it is
- VFS dir updated to accept both entries as:
   - if it gets a file it will forget its parent dir + subdirs
   - if it gets a folder it will behave the same as before
- fstests updated and regenerated
   - the test was updated too, it now checks for both a file and a folder to be received in their full paths
- cache updated to use new polling mechanism

Don't merge it yet, even if tests are successful. I would like to do some testing on ACD

remusb commented:
I'll squash the commits in the PR after this

remusb commented:
Hmm, weird failure reason. 
```
go list ./... | xargs -n1 golint | grep -E -v '(StorageUrl|CdnUrl)' ; test $? -eq 1
/Users/travis/gopath/src/github.com/ncw/rclone/fs/walk/walk.go:106:9: should omit values from range; this loop is equivalent to `for range ...`
make: *** [check] Error 1
```

That code isn't covered in this PR and other recent tests don't have this problem.

remusb commented:
And hung too apparently. I restarted it.

ncw commented:
I fixed the golint warning in e0eb666dbf415ad16826c647e9a986bb9722e76c - it is a new test.

Not sure why we are getting the hangs - travis seems to be getting unreliable :-(

remusb commented:
Everything looks fine in the tests so far but sadly I can't do a regression on ACD which is the only other remote that uses these interfaces.

Drive works though. If the PR and changes look good to you @ncw I'm ok with pushing these in master. It should lock the whole push - notify logic I wanted for cache

ncw commented on backend/cache/cache.go:
```diff
@@ -381,16 +385,16 @@ func NewFs(name, rootPath string) (fs.Fs, error) {
 	}).Fill(f).Mask(wrappedFs).WrapsFs(f, wrappedFs)
 	// override only those features that use a temp fs and it doesn't support them
-	f.features.DirChangeNotify = f.DirChangeNotify
+	//f.features.ChangeNotify = f.ChangeNotify
```
Is this intended?  It looks like cache no longer supports ChangeNotify?

remusb replied:
It just supports it when the wrapped fs does too. Cache doesn't provide notifications on its own and tests (at the moment only those) fail when it wraps a remote that doesn't support it.

Perhaps it will in the future, regardless of the wrapped remote, but at the moment to just add notifications on every change can likely trigger a notification mess so better to play it safe.

ncw replied:
Oh, I see!

ncw commented on backend/drive/drive.go:
```diff
@@ -1237,28 +1234,35 @@ func (f *Fs) dirchangeNotifyRunner(notifyFunc func(string), pollInterval time.Du
 		}
 
-		pathsToClear := make([]string, 0)
+		pathsToClear := make(map[string]fs.EntryType)
```
Using a map instead of a slice has the potential to get them out of order.  Do you think this will matter?  If we want to preserve the order then easy enough to use `[]struct{string, fs.EntryType}`

remusb replied:
I don't believe it matters. The previous implementation was sorting them anyway: `sort.Strings(pathsToClear)` so that broke the order it got them from the cloud provider.

ncw replied:
I guess I was thinking ahead if we were to be slightly more specific in the notifications - this file deleted, this directory renamed for which the order would be more important.

remusb replied:
@ncw Fair enough. That's a good point. I switched to an array of struct to keep the order and adjusted the test to verify that the order is kept too.

ncw commented:
I think you should merge now so we can get some testing in before the release :-)

remusb commented:
I updated to maintain the order. I'm going to merge it in after you take another look at the changes. I want to make sure that the funky stuff with the struct type in the function is ok with you.
@ncw

ncw commented:
> I updated to maintain the order. I'm going to merge it in after you take another look at the changes. I want to make sure that the funky stuff with the struct type in the function is ok with you.

I gave the code the once over.  It looks fine to me :-)

Can you squash this branch a bit before to remove the false steps from the commit history?

remusb commented:
Seems like a transient issue or some sort. Is there a way to retest on AppVeyor? Is it a matter of permissions? @ncw

ncw commented:
> Seems like a transient issue or some sort. Is there a way to retest on AppVeyor? Is it a matter of permissions?

I've given it a kick.  You should be able to log in to appveyor - use your github user to log in - then you should have permissions to do that too I think.

remusb commented:
@ncw sadly no, even logged in I don't have any option to retry. On Travis I do though.

ncw commented:
@remusb I think I clicked enough things that you should have permission now.  [The docs](https://www.appveyor.com/docs/team-setup/) are rather confusing!  Luckily I found [the forum](https://help.appveyor.com/discussions/questions/569-allow-github-project-collaborator-to-configure-appveyor-project-configuration).

remusb commented:
@ncw I can see the repo and all the builds, just not do anything with them. Only download the log from what . I can see.
Don't both that much with it though.
